Port kernel-install to /bin/sh

nabijaczleweli opened the pull request:
All relatively trivial and mechanical, save for the occasional in(s)ane bash extension dances, those took some untangling

Supersedes #11751

nabijaczleweli commented on src/kernel-install/90-loaderentry.install:
```diff
@@ -106,41 +91,35 @@ install -g root -o root -m 0644 "$KERNEL_IMAGE" "$ENTRY_DIR_ABS/linux" || {
-INITRD_OPTIONS=( "${@:${INITRD_OPTIONS_START}}" )
+shift "$INITRD_OPTIONS_SHIFT"
+for initrd; do
+    [ -f "$initrd" ] || continue
```
BTW, I think this is a behaviour bug. If an *explicitly-supplied* initrd doesn't exist it's an error, and shouldn't get eaten.

behrmann commented on src/kernel-install/kernel-install:
```diff
@@ -182,32 +169,22 @@ case $COMMAND in
-            fi
+        for f in $PLUGINS; do
+            [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && echo "+$f add $KERNEL_VERSION $ENTRY_DIR_ABS $KERNEL_IMAGE $*"
+            "$f" add "$KERNEL_VERSION" "$ENTRY_DIR_ABS" "$KERNEL_IMAGE" "$@"
```
I think losing `INITRD_OPTIONS` is an unnecessary loss of leggibility.

nabijaczleweli replied:
Whereas I can read this just fine! Amazing what fluency in a language does!

behrmann replied:
Fluency is a wonderful thing, but code is also about communicating intent to other people and the next person looking at this might not well-versed in POSIX shell or be in a hurry. There's a reason to give variables sensible names. In this case one has to follow the whole script to follow mutations of `$@` throughout. A simple copy earlier on would make it easy to see at a glance what this is.

nabijaczleweli replied:
If you aren't well-versed in shell then you shouldn't be writing a program in shell without testing it throughout. Also, there's literal 300 lines of XML describing what this line does in intricate detail. Either way, this is a made-up problem.

keszybz replied:
I'd prefer to keep INITRD_OPTIONS too. It does make the code more readable. And no, we can't assume that every person who reads this code or even changes it is fluent in shell. Some complexity and required implicit knowledge cannot be avoided, but we should avoid that wherever possible.

behrmann reviewed (commented):
it's very nice, that this PR makes `kernel-install` and friends shellcheck clean.

There's a bit of churn in there because of things that might only be stylistic changes (`-r` vs `-f`) and I looking at some of the contortions I would challenge the opinion, that bash extensions are insane :)

nabijaczleweli commented on src/kernel-install/50-depmod.install:
```diff
@@ -20,23 +20,25 @@
 
 COMMAND="$1"
 KERNEL_VERSION="$2"
-ENTRY_DIR_ABS="$3"
-KERNEL_IMAGE="$4"
-INITRD_OPTIONS_START="5"
```
They're also unused and the documentation is provided in the documentation. More important documentation is that 50-depmod only uses $1 and $2

nabijaczleweli commented on src/kernel-install/50-depmod.install:
```diff
@@ -20,23 +20,25 @@
-KERNEL_IMAGE="$4"
-INITRD_OPTIONS_START="5"
-
-[[ $KERNEL_VERSION ]] || exit 1
```
Already guaranteed in k-i L79:
```sh
if [ -z "$COMMAND" ] || [ -z "$KERNEL_VERSION" ]; then
```

nabijaczleweli commented on src/kernel-install/90-loaderentry.install:
```diff
@@ -22,68 +22,53 @@ COMMAND="$1"
-fi
+case "$COMMAND" in
+    remove)
+        exec rm -f \
+            "$BOOT_ROOT/loader/entries/$MACHINE_ID-$KERNEL_VERSION.conf" \
+            "$BOOT_ROOT/loader/entries/$MACHINE_ID-$KERNEL_VERSION+"*".conf"
```
It doesn't. It would. It does. They are.

behrmann replied:
Can you guarantee that? :) 

A quick look at coreutils tells me that `-f` only makes `rm` non-interactive, disables prompts and ignores missing files, besides, the actual `rm` function doesn't even know it is being forced. There's plenty of other sources of error though and a quick test case shows us:
```
$ touch foo
$ chattr +i foo 
$ rm -f foo 
rm: cannot remove 'foo': Operation not permitted
$ echo $?
1
```
So your assertion doesn't hold. While this example is artificial, there might be other reasons for `rm` to fail, e.g. a full filesystem.

nabijaczleweli replied:
Yes, and in that case 90- should fail because it failed to actually remove.

behrmann replied:
Sure, this might be the correct behaviour, I haven't thought about this yet and happily leave this to others, I just wanted to highlight that in the first place this *is* a behavioural change.

keszybz replied:
I think this change is fine, but please mention it in the commit message.

nabijaczleweli replied:
done

nabijaczleweli commented on src/kernel-install/kernel-install:
```diff
@@ -18,38 +18,31 @@
 # along with systemd; If not, see <http://www.gnu.org/licenses/>.
 
-SKIP_REMAINING=77
+skip_remaining=77
```
Because it's not a global in the actual sense (this changes some shellcheck semantics, and reads better because it doesn't originate from the environment).

behrmann replied:
Yes, but this is a constant, which are traditionally uppercased.

nabijaczleweli replied:
It's a variable that originates locally and not from the environment => it's lcase. You'll do well to note /bin/sh doesn't have constants.

keszybz commented:
Hmm, before I review this: why would we want to do this? We don't program in C89, and we don't need to posix ourselves harder with plain sh.

nabijaczleweli commented:
In your analogy, /bin/sh is to bash not as C89 to C11, but as C11 to Microsoft Visual C++ 6 with all your favourite extensions. Both /bin/sh and C11 are actively maintained and kept up-to-date (you may have noticed this from `command -v` making it to the base spec, or from most of the good Almquist extensions doing so over the years).

Bash is great as a funny scripting language for your builds with layering violations every third line, but nevertheless it's not required to be a part of a base system. Would you say the same thing if your bootloader required Tcl (okay, extremist example), or any other shell dialect extension variant (zsh, or, more realistically, the KornShell) that you liked less? Or that wasn't part of the fedora Essential set like bash is for some reason?

/bin/sh is better-understood, better-defined, and it came out shorter and simpler despite inlining the big brace expansion in 50-.

behrmann commented:
> /bin/sh is better-understood, better-defined, and it came out shorter and simpler despite inlining the big brace expansion in 50-.

Well, shorter yes, simpler - under some metric - maybe, but I'll contest easier to understand; simple things are not necessarily easy or easier to understand. 

I'd also argue that POSIX shell might be less well understood than its adherents argue, since large(r) shell projects, e.g. some that produce relevant outputs for kernel-install like dracut or mkinitcpio, are written in bash instead of POSIX shell and I'd hazard that this is rather the rule than the exception, i.e. more people write more scripts in bash than POSIX shell and more people use bash as their interactive shell, which is why I'd argue that more people are used to bash's features and idiosyncrasies than that of POSIX shell.

nabijaczleweli commented:
Hm, also, and this may prove more convincing, this is the last bit of the distribution that hard-depends on bash. There's two other uses: in `nspawn/nspawn.c`, but that first tries bash then sh as login shells (this is fine), and `shared/user-record.c` which returns `/bin/bash` by default for regular users (no fallback, but easy enough to add one with `return access("/bin/bash", X_OK) ? "/bin/sh" : "/bin/bash";` instead).

This, coupled with us not losing anything from standardising on /bin/sh here, means bash could be dropped as a hard dependency, which is quite pogged.
Dropping bash (well, not having the capability of having bash on some systems) is also cited in the original #11751.

keszybz commented on src/kernel-install/50-depmod.install:
```diff
@@ -20,23 +20,25 @@
+            "/lib/modules/$KERNEL_VERSION/modules.devname" \
+            "/lib/modules/$KERNEL_VERSION/modules.softdep" \
+            "/lib/modules/$KERNEL_VERSION/modules.symbols" \
+            "/lib/modules/$KERNEL_VERSION/modules.symbols.bin"
```
OK, this **is** more readable.

keszybz commented on src/kernel-install/90-loaderentry.install:
```diff
@@ -22,68 +22,53 @@ COMMAND="$1"
-elif [[ -f /usr/lib/kernel/cmdline ]]; then
-    read -r -d '' -a BOOT_OPTIONS < /usr/lib/kernel/cmdline
+if [ -r /etc/kernel/cmdline ]; then
+    BOOT_OPTIONS="$(tr -s "$IFS" ' ' < /etc/kernel/cmdline)"
```
Could we get rid of the space after `<` while at it?

nabijaczleweli replied:
Sure, trimmed all spaces in redirects (`git grep '[<>]\s'` is empty)

keszybz commented on src/kernel-install/kernel-install:
```diff
@@ -182,7 +180,7 @@ case "$COMMAND" in
             "$f" remove "$KERNEL_VERSION" "$ENTRY_DIR_ABS"
             err=$?
-            [ $err -eq $skip_remaining ] && break
+            [ $err -eq 77 ] && break
```
Hmm, I think it was better to have the variable name as documentation.

nabijaczleweli replied:
Sure, dropped

keszybz commented on src/kernel-install/90-loaderentry.install:
```diff
@@ -106,40 +91,35 @@ install -g root -o root -m 0644 "$KERNEL_IMAGE" "$ENTRY_DIR_ABS/linux" || {
     exit 1
 }
 
-INITRD_OPTIONS=( "${@:${INITRD_OPTIONS_START}}" )
```
Here too, I think INITRD_OPTIONS should stay.

keszybz reviewed (commented):
Thanks, this looks pretty nice.

nabijaczleweli commented:
Updated, rebased, all applied except as noted below.

In Re: `INITRD_OPTIONS` in k-i: that's an outdated diff (you can tell due to the big orange Outdated badge), and has been for some time now – the funny names masked the fact that everything since the *image* is part of the passthrough bundle, and nothing actually uses the image. See up-to-date L173.

In Re: `INITRD_OPTIONS` in 90-: what? how? I guess I can put a comment on the shift, but there's a huge `INITRD_OPTIONS_SHIFT` rag, multiple times, and all the `for`s call the variable `initrd`, so a comment on the I_O_S definition?
